java/src/jni/h5util.c

```diff
@@ -210,9 +210,9 @@ h5str_convert(JNIEnv *env, char **in_str, hid_t container, hid_t tid, void *out_
     /* Build default formats for long long types */
     if (!fmt_llong[0]) {
         if (HDsprintf(fmt_llong, "%%%sd", H5_PRINTF_LL_WIDTH) < 0)
-            H5_JNI_FATAL_ERROR(ENVONLY, "h5str_convert: HDsprintf failure");
+            H5_ASSERTION_ERROR(ENVONLY, "h5str_convert: HDsprintf failure");
```

Comment:
Again, sprintf, snprintf, fprintf, fopen and related should remain fatal errors because they indicate a system error that occurred when trying to output something. I don't think an assertion error is the most appropriate to use for h5str_append either, but if we're going to use it, only make the changes for h5str_append. Don't change anything else, because that's not what assertions are for.

Assertions are for code that we expect not to execute or be TRUE, because we've explicitly written it that way. For example, asserting after a failure from HDfopen makes no sense. We didn't expect the failure.

Reply:
I disagree, the assertion is that the function worked properly and I think the assertion error is closer to what we want, again why crash the JVM if a string op failed.

Reply:
It's not about agreeing or disagreeing, this is literally what assertions are used for. To assert that some condition you expect to be true is actually true; they work essentially the same way as in C.

The reason to throw a fatal error when the printf family of functions fail is because it is exceedingly rare and would generally indicate an encoding error or even a system error.

Reply:
Okay whatever!

Reply:
Let's just change just the h5str_append calls to use the new assertion error and leave it at that. The more changes we make here, the wider the scope of the PR, which doesn't help getting the main changes in here.